alts/src/main/java/io/grpc/alts/ComputeEngineChannelCredentials.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.alts;
+
+import com.google.auth.oauth2.ComputeEngineCredentials;
+import com.google.common.collect.ImmutableList;
+import io.grpc.CallCredentials;
+import io.grpc.ChannelCredentials;
+import io.grpc.CompositeChannelCredentials;
+import io.grpc.ExperimentalApi;
+import io.grpc.Status;
+import io.grpc.alts.internal.AltsProtocolNegotiator.GoogleDefaultProtocolNegotiatorFactory;
+import io.grpc.auth.MoreCallCredentials;
+import io.grpc.internal.SharedResourcePool;
+import io.grpc.netty.GrpcSslContexts;
+import io.grpc.netty.InternalNettyChannelCredentials;
+import io.grpc.netty.InternalProtocolNegotiator;
+import io.netty.handler.ssl.SslContext;
+import javax.net.ssl.SSLException;
+
+/**
+ * Credentials for Google Compute Engine. This class sets up a secure channel using ALTS if
+ * applicable and using TLS as fallback.
+ */
+@ExperimentalApi("TODO")
+public final class ComputeEngineChannelCredentials {
+  private ComputeEngineChannelCredentials() {}
+
+  /**
+   * Creates credentials for Google Compute Engine. This class sets up a secure channel using ALTS
+   * if applicable and using TLS as fallback.
+   */
+  public static ChannelCredentials create() {
+    ChannelCredentials googleDefaultNegotiator =
+        InternalNettyChannelCredentials.create(createClientFactory());
+    CallCredentials callCredentials;
+    if (CheckGcpEnvironment.isOnGcp()) {
+      callCredentials = MoreCallCredentials.from(ComputeEngineCredentials.create());
+    } else {
+      callCredentials = new FailingCallCredentials(
```

Comment:
Have been meaning to ask this. I understand this matches the old behavior but if you are not on Gcp is there a reason to fail each call instead of failing channel creation?  Wouldn't it be better to fail early when the environment is wrong?

Reply:
We don't tend to fail channel creation based on I/O; we do it when you use the API incorrectly. So I think we did it this way to fail in the normal way we fail for I/O.